src/plugin_host.jl

```diff
@@ -0,0 +1,174 @@
+    else
+        proc = require_plug(h, method)
+
+        if proc == nothing
+            reply_error(c, serial, "Command not defined.")
+        end
+
+        @async(try
+            reply_result(c, serial, proc(c, args...))
+        catch
+            err = "No matching method for args (NvimClient, " * join(args, ", ") * ")"
+            reply_error(c, serial, err)
+        end)
+    end
+end
+
+function require_plug(h::HostHandler, name::String)
+    (plugin_file, proc_id) = split(name, ':', 2)
+
+    if plugin_file ∉ h.loaded_plugins
+        cbs = get_callbacks(plugin_file)
+        merge!(h.proc_callbacks, cbs)
+    end
+
+    get(h.proc_callbacks, name, nothing)
+end
+
+function get_specs(plugin_file)
+    plugin = nothing
+    try
+        plugin = decorate(plugin_file)
+    end
+
+    fn_specs = {}
+    global plug(proc_type, name, handler, opt_args...) = begin
+        conf = Dict{String, Any}()
+        conf["type"] = proc_type
+        conf["name"] = name
+        opts = Dict{String, Any}()
+        for (opt_k, opt_v) in opt_args
+            opts[string(opt_k)] = opt_v
+        end
+        conf["opts"] = opts
+        conf["sync"] = pop!(opts, "sync", 0)
+        push!(fn_specs, conf)
+    end
+
+    try
+        eval(plugin)
+    catch err
+        println(STDERR, "Error while loading plugin " * plugin_file)
+        println(STDERR, err)
+    end
+
+    fn_specs
+end
+
+function get_callbacks(plugin_file)
+    plugin = nothing
+    try
+        plugin = decorate(plugin_file)
+    end
+
+    proc_callbacks = Array((String, Expr), 0)
+    global plug(proc_type, name, handler, opt_args...) = begin
+        pattern = ""
+        for arg in opt_args
+            if arg[1] == "pattern"
+                pattern = ":" * arg[2]
+            end
+        end
+
+        proc_name = "$plugin_file:$proc_type:$name$pattern"
+        push!(proc_callbacks, (proc_name, parse(handler)))
+    end
+
+    try
+        eval(plugin)
+    catch err
+        println(STDERR, "Error while loading plugin " * plugin_file)
+        println(STDERR, err)
+    end
+
+    extract_fns(x) = (x[1], eval(x[2]))
+    Dict{String, Function}(map(extract_fns, proc_callbacks))
+end
+
+# finds constructs that look like decorators (i.e. macrocall then function)
+# and then adds the function name as the last argument to the macrocall
+function decorate(file_name::String)
+    ast = (open(readall, file_name) |> parse)::Expr
+    module_name = ast.args[2]
+
+    q = Array(Expr, 0)
+    candidate_fns = Array((Expr, Expr), 0)
+    push!(q, ast)
+    while length(q) > 0
+        expr = shift!(q)
+        for (i, arg) in enumerate(expr.args)
+            if typeof(arg) != Expr continue end
+            if arg.head == :macrocall && expr.args[i + 2].head == :function
+                # `i + 2` to skip LineNumberNode
+                fn_name = expr.args[i + 2].args[1].args[1]
+                push!(arg.args, symbol("$module_name.$fn_name"))
```

Comment:
If you're willing to drop support for Julia v0.3, then the new doc syntax will actually work:
Just replace the current macro scheme with a single call to `@Neovim.plugin` that puts the required module's name in `task_local_storage`.
Next, use the following function
```julia
function get_cmds(plugin::Module)
    plugin_cmds = []
    for f in filter(k->isa(k, Function), keys(plugin.META))
        spec = Base.doc(f) |> Base.Markdown.plain |> parse
        push!(plugin_cmds, (f, spec))
    end
    return plugin_cmds
end
```
and then process the Option 2 function call.

While `get_cmds` may seem ad-hoc, the default doc syntax is markdown, so plain-stringifying is within spec. There's still this sort of two-pass thing going on, but it avoids the problems of using macros.